crates/turborepo-filewatch/src/lib.rs

```diff
@@ -89,53 +92,96 @@ pub struct FileSystemWatcher {
+
+        if root.relation_to_path(&cookie_dir) != PathRelation::Parent {
             return Err(WatchError::Setup(format!(
                 "Invalid cookie directory: {} does not contain {}",
                 root, cookie_dir
             )));
         }
-        setup_cookie_dir(cookie_dir)?;
-        let (sender, _) = broadcast::channel(1024);
+
+        let (sender_tx, sender_rx) = OptionalWatch::new();
+        let (receiver_tx, receiver_rx) = OptionalWatch::new();
         let (send_file_events, mut recv_file_events) = mpsc::channel(1024);
-        let watch_root = root.to_owned();
-        let broadcast_sender = sender.clone();
-        debug!("starting filewatcher");
-        let watcher = run_watcher(&watch_root, send_file_events)?;
         let (exit_ch, exit_signal) = tokio::sync::oneshot::channel();
-        // Ensure we are ready to receive new events, not events for existing state
-        debug!("waiting for initial filesystem cookie");
-        wait_for_cookie(cookie_dir, &mut recv_file_events).await?;
-        tokio::task::spawn(watch_events(
-            watcher,
-            watch_root,
-            recv_file_events,
-            exit_signal,
-            broadcast_sender,
-        ));
-        debug!("filewatching ready");
+
+        tokio::task::spawn({
+            let cookie_dir = cookie_dir.clone();
+            let watch_root = root.to_owned();
+            async move {
+                // this task never yields, so run it in the blocking threadpool
+                let watch_root_task = watch_root.clone();
+                let cookie_dir_task = cookie_dir.clone();
+                let task = tokio::task::spawn_blocking(move || {
+                    setup_cookie_dir(&cookie_dir_task)?;
+                    run_watcher(&watch_root_task, send_file_events)
+                });
+
+                let Ok(Ok(watcher)) = task.await else {
+                    // if the watcher fails, just return. we don't set the event sender, and other
+                    // services will never start
+                    return;
+                };
+
+                // Ensure we are ready to receive new events, not events for existing state
+                debug!("waiting for initial filesystem cookie");
+                if let Err(e) = wait_for_cookie(&cookie_dir, &mut recv_file_events).await {
+                    // if we can't get a cookie here, we should not make the file
+                    // watching available to downstream services
+                    warn!("failed to wait for initial filesystem cookie: {}", e);
+                    return;
+                }
+                debug!("filewatching ready");
+
+                let (sender, receiver) = broadcast::channel(1024);
+
+                // if either of these things fail, it means that events that we either won't
+                // receive events or don't have any downstream listeners. In either case, we
+                // should probably not run the watcher
```

Comment:
The comment implies we should care about the errors, but we're ignoring them. Also, I think we still want filewatching even if we don't yet have any downstream consumers.

Reply:
Updated with some better explanation